### PR TITLE
[8.0] Getting more details about failed/aborted pilots from HTCondor

### PR DESCRIPTION
**Pull request by aldbr**
This PR:
- allows operators to get output data even if the pilots fail/are aborted:
  - by default, HTCondor does not mark pilots as `Failed` if they don't terminate correctly.
  - the `on_exit_hold` option allows to mark pilots as `held` if they fail (`on_exit_hold_subcode` allows to identify them with a specific code).
  - from there, the output data can be transferred if needed, and DIRAC marked them as `Failed`.
  - the `periodic_remove` deletes the pilots at some point.

- fixes wrong `Aborted` pilots:
  - some pilots are marked as `Aborted` in  DIRAC while there are temporary `Held` in HTCondor because inputs are spooled.
  - these pilots are now marked as `Waiting`.

BEGINRELEASENOTES
*Resources
FIX: getting more details about failed/aborted pilots from HTCondor 
ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
Please fix the conflict.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 1]
```diff
@@ -152,11 +108,49 @@ def __init__(self, ceUniqueID):
         self.tokenFile = None
 
     #############################################################################
-    def __writeSub(self, executable, nJobs, location, processors, pilotStamps, tokenFile=None):
+
+    def _DiracToCondorID(self, diracJobID):
+        """Convert a DIRAC jobID into an Condor jobID.
+        Example: https://<ce>/1234/0 becomes 1234.0
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]:
```suggestion
        Example: https://<ce>/1234.0 becomes 1234.0
```

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 2]
```diff
@@ -152,11 +108,49 @@ def __init__(self, ceUniqueID):
         self.tokenFile = None
 
     #############################################################################
-    def __writeSub(self, executable, nJobs, location, processors, pilotStamps, tokenFile=None):
+
+    def _DiracToCondorID(self, diracJobID):
+        """Convert a DIRAC jobID into an Condor jobID.
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]:
```suggestion
        """Convert a DIRAC jobID into an Condor jobID.
```
The use of DIRAC jobID is confusing here, this has nothing to do with a DIRAC job, but rather a pilot identifier?
So it turns a pilot Identifier into a condor jobID

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 3]
```diff
@@ -152,11 +108,49 @@ def __init__(self, ceUniqueID):
...
+        clusterIDs = [clu.strip() for clu in clusterIDs]
+        self.log.verbose("Cluster IDs parsed:", clusterIDs)
+        try:
+            clusterID = clusterIDs[0].split(".")[0]
+            numJobs = clusterIDs[1].split(".")[1]
+        except IndexError:
+            return S_ERROR(f"Something wrong with the condor_submit output: {condorJobIDs}")
+
+        cePrefix = f"htcondorce://{self.ceName}/"
+        jobReferences = [f"{cePrefix}{clusterID}.{i}" for i in range(int(numJobs) + 1)]
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:145]: This has the dot in the pilot identifier

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 4]
```diff
@@ -19,50 +13,127 @@
...
+
+Queue stamp in %(pilotStampList)s
+"""
+
+
 def parseCondorStatus(lines, jobID):
     """parse the condor_q or condor_history output for the job status
 
-    :param lines: list of lines from the output of the condor commands, each line is a pair of jobID and statusID
+    :param lines: list of lines from the output of the condor commands, each line is a pair of jobID, statusID, and holdReasonCode
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py]:
```suggestion
    :param lines: list of lines from the output of the condor commands, each line is a tuple of jobID, statusID, and holdReasonCode
```

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 5]
```diff
@@ -19,50 +13,127 @@
...
-    status = sp.returncode
-
-    # Join the ClusterId and the ProcId and add to existing list of statuses
-    if status == 0:
-        for line in output.split("\n"):
-            values = line.strip().split()
-            if len(values) == 3:
-                qList.append("%s.%s %s" % tuple(values))
+        if l[0] == jobID:
+            # A job can be held for many various reasons, we need to further investigate with the holdReasonCode & holdReasonSubCode
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py]:
```suggestion
            # A job can be held for various reasons, we need to further investigate with the holdReasonCode & holdReasonSubCode
```

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 6]
```diff
@@ -183,55 +177,26 @@ def __writeSub(self, executable, nJobs, location, processors, pilotStamps, token
                 """
             )
 
+        # Remote schedd options by default
+        targetUniverse = "vanilla"
         # This is used to remove outputs from the remote schedd
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: This comment is now sitting in the wrong place, this belongs to the periodic remove

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 7]
```diff
@@ -291,15 +256,28 @@ def _executeCondorCommand(self, cmd, keepTokenFile=False):
...
+            self.tokenFile = None
+            # We have got a non-zero status code
+            errorString = stderr if stderr else stdout
+            return S_ERROR(f"Command {cmd} failed with: {status} - {errorString.strip()}")
+
         # Remove token file if we do not want to keep it
         self.tokenFile = self.tokenFile if keepTokenFile else None
 
-        return result
+        return S_OK(stdout.strip())
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:279]: So the function no longer returns the result of the executeGridCommand, please update the docstring.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 8]
```diff
@@ -377,27 +345,20 @@ def killJob(self, jobIDList):
             jobIDList = [jobIDList]
 
         self.log.verbose("KillJob jobIDList", jobIDList)
-
         self.tokenFile = None
 
-        for jobRef in jobIDList:
-            job, _, jobID = condorIDAndPathToResultFromJobRef(jobRef)
-            self.log.verbose("Killing pilot", job)
+        for diracJobID in jobIDList:
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Why change this to diracJobID, when these are not actually dirac job IDs?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 9]
```diff
@@ -443,77 +404,56 @@ def getJobStatus(self, jobIDList):
         resultDict = {}
         condorIDs = {}
         # Get all condorIDs so we can just call condor_q and condor_history once
-        for jobRef in jobIDList:
-            job, _, jobID = condorIDAndPathToResultFromJobRef(jobRef)
-            condorIDs[job] = jobID
+        for diracJobID in jobIDList:
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]:
```suggestion
        condorIDs = dict({(pR := jobRef.split(":::")[0], self._DiracToCondorID(pR)) for jobRef in jobIDList})
```
Fixed to make a dict.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 10]
```diff
@@ -443,77 +404,56 @@ def getJobStatus(self, jobIDList):
...
             condorHistCall.extend(_condorIDs)
-            condorHistCall.extend(["-af", "ClusterId", "ProcId", "JobStatus"])
+            condorHistCall.extend(["-af:j", "JobStatus", "HoldReasonCode", "HoldReasonSubCode", "HoldReason"])
+            result = self._executeCondorCommand(cmd, keepTokenFile=True)
+            if not result["OK"]:
+                return result
 
-            self._treatCondorHistory(condorHistCall, qList)
+            _qList = result["Value"].split("\n")
+            qList.extend(_qList)
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]:
```suggestion
            qList.extend(result["Value"].split("\n"))
```

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 11]
```diff
@@ -443,77 +404,56 @@ def getJobStatus(self, jobIDList):
...
-                return S_ERROR(f"condor_q failed completely: {result['Message']}")
-            status, stdout, stderr = result["Value"]
-            if status != 0:
-                self.tokenFile = None
-                return S_ERROR(stdout + stderr)
-            _qList = stdout.strip().split("\n")
+                return result
+
+            _qList = result["Value"].split("\n")
             qList.extend(_qList)
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]:
```suggestion
            qList.extend(result["Value"].split("\n"))
```

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 12]
```diff
@@ -443,77 +404,56 @@ def getJobStatus(self, jobIDList):
...
         for job, jobID in condorIDs.items():
-            pilotStatus = parseCondorStatus(qList, jobID)
-            if pilotStatus == "HELD":
-                # make sure the pilot stays dead and gets taken out of the condor_q
-                cmd = f"condor_rm {self.remoteScheddOptions} {jobID}".split()
-                _result = self._executeCondorCommand(cmd, keepTokenFile=True)
-                pilotStatus = PilotStatus.ABORTED
+            pilotStatus, reason = parseCondorStatus(qList, jobID)
+
+            if pilotStatus == PilotStatus.ABORTED:
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Is this going to cancel jobs over and over again? Before the status was `HELD` once and then set to ABORTED, now if they are cancelled, this is ABORTED, and might remain so, so killJob is called for them until deleted from the system? As the list also uses information from condor_history?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 13]
```diff
@@ -540,13 +480,39 @@ def getJobOutput(self, jobID):
 
         return S_OK((result["Value"]["output"], result["Value"]["error"]))
 
+    def _findFile(self, workingDir, fileName, pathToResult):
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Why this change?
This function doesn't use self. It could use self.workingDirectory instead of as an argument probably.
Bot now it just makes things harder to review.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 14]
```diff
@@ -19,50 +13,127 @@
...
+
+# Directory that will contain the outputs
+initialdir = %(initialDir)s
+
+# Outputs: stdout, stderr, log
+output = $(Cluster).$(Process).out
+error = $(Cluster).$(Process).err
+log = $(Cluster).$(Process).log
+
+# Transfer all output files, even if the job is failed
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py]: as far as I remember the point of `transfer_output_files = ""` is to not transfer all output files, but just logs (#3447), so the comment needs clarification.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 15]
```diff
@@ -19,50 +13,127 @@
...
+# ------------
+# Specify the signal sent to the job when HTCondor needs to vacate the worker node
+kill_sig=SIGTERM
+# By default, HTCondor marked jobs as completed regardless of its status
+# This option allows to mark jobs as Held if they don't finish successfully
+on_exit_hold = ExitCode != 0
+# A random subcode to identify who put the job on hold
+on_exit_hold_subcode = %(holdReasonSubcode)s
+# Jobs are then deleted from the system after N days if they are not running
+period_remove = (JobStatus != 1) && (JobStatus != 2) && ((time() - EnteredCurrentStatus) > (%(daysToKeepRemoteLogs)s * 24 * 3600))
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py]: This was a remote_schedd only option

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 16]
```diff
@@ -19,50 +13,127 @@
...
+
+# Outputs: stdout, stderr, log
+output = $(Cluster).$(Process).out
+error = $(Cluster).$(Process).err
+log = $(Cluster).$(Process).log
+
+# Transfer all output files, even if the job is failed
+transfer_output_files = ""
+should_transfer_files = YES
+when_to_transfer_output = ON_EXIT_OR_EVICT
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py:53]: This were local_schedd options.
Actually the spelling was different (CamelCase), and I don't know what is correct.
https://htcondor.readthedocs.io/en/lts/grid-computing/grid-universe.html#htcondor-c-job-submission

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 2 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: You are right, the name is not correct but here is an explanation. I wanted to bring more homogeneity between the `AREX` and the `HTCondor` CE interfaces. We use the `AREXCE` to submit pilot-jobs but also jobs. Theoretically, we could also use the `HTCondorCE` to submit jobs, which is why I did not want to call this a pilot identifier.

But indeed, it is neither a dirac job ID nor a dirac pilot reference. I would like to find a generic name for that (`genericJobID`? `jobID` - as opposed to `condorJobID`? `jobReference`?). Any opinion? Suggestion?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 3 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:145]: Not sure to understand your comment. Is there anything wrong here?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 9 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: I did not implement this: it creates a set of tuples.
But I shorten the number of lines in the loop if it helps.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 12 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: I forgot to remove the line that calls `killJob()`.
The idea is that held jobs would be automatically removed by the `periodic_remove` condition.
Any opinion?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 13 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: I included `self.workingDirectory` within the method. Is that better?

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 15 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py]: This is part of the logic to delete the held jobs now.
Would it be a problem with the local_schedd?

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 16 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py:53]: We moved them here because it should help us getting the outputs of the aborted pilots.
I guess we were using the vanilla options: https://htcondor.readthedocs.io/en/latest/users-manual/file-transfer.html?#specifying-what-files-to-transfer

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 3 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:145]: Just a cross-reference to the slash that was in the docstring before.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 13 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Not better with respect to reviewing.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 16 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py:53]: It "should" or it does?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 9 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: I wanted to propose something, where we don't have to say diracJobID.
I fixed the comprehension to make a dict at the end.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 15 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py]: Should be OK

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 2 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: > We use the AREXCE to submit pilot-jobs but also jobs.

I had no idea about this possibility. What does this thing look like in case this is a "job"?

Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: > I had no idea about this possibility. What does this thing look like in case this is a "job"?

We use the `AREXCE` to push an application (e.g. a bash script containing `<command> <arguments>`) from a workflow module and wait for its termination, as part of the `PushJobAgent` execution (to execute jobs in computing resources with no external connectivity).
In the future, we would like to directly push the `dirac-jobexec` command using the `AREXCE`.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 16 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py:53]: I checked again because I was not entirely sure with all the changes I made: it does if a pilot is removed.
I made a test by removing a `FAILED` pilot with `condor_rm`:
- `when_to_transfer_output = ON_EXIT_OR_EVICT`: I get the outputs
- `when_to_transfer_output = ON_EXIT`: 

```bash
Failed to find output for condor job htcondorce...
{'OK': False, 'Errno': 2, 'Message': 'No such file or directory ( 2 : Could not find ...'}
```

Comment (andresailer) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py:53]: Ok, good!

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 2 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Thanks for the explanation!
What does `diracJobID` look like in that case?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 12 above)
Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: This should be OK. at the moment I still have a periodic_remove in my extraOptions thing, but with the fixes you have in here, I probably don't need that any more.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 2 above)
Comment (aldbr) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: There is no difference, the `diracJobID` is the same in both cases, namely a transformation of the CE-specific ID to a "generic" format: `1234` becomes `<protocol>://<cename + extra fields like port>/1234`.
Then:
- if it is a pilot-job, it is interpreted by the caller of the CE (the `SiteDirector`) as a pilot job reference.
- if it is an application, it is interpreted by the caller of the CE (a Workflow module) as a "job-application" ID let's say (it is just named `jobID` in the code, you can see that in `RemoteRunner`).

Does this answer to your question?

Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py]: Yeah, so calling this a jobReference is probably best?

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 17]
```diff
@@ -19,50 +13,137 @@
...
+request_cpus = %(processors)s
+
+# Exit options
+# ------------
+# Specify the signal sent to the job when HTCondor needs to vacate the worker node
+kill_sig=SIGTERM
+# By default, HTCondor marked jobs as completed regardless of its status
+# This option allows to mark jobs as Held if they don't finish successfully
+on_exit_hold = ExitCode != 0
+# A random subcode to identify who put the job on hold
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py]:
```suggestion
# A subcode of our choice to identify who put the job on hold
```

**Review by andresailer — DISMISSED** [review]
Sorry you were waiting for me for so long.

Emoji responses, and marking discussions as resolved don't trigger notifications.

So I was expecting explicit responses to comments with "Done", or a final: "I changed things along the lines you proposed" to have another look.

**Comment by aldbr** [comment]
Indeed, my fault.
Next time, I will comment or click on the "re-request review" button.

Thanks.

**Review by fstagni — APPROVED** [review]
